src/backend/pipeline/cont_adhoc_sender.c

```diff
@@ -0,0 +1,373 @@
+#include "postgres.h"
+#include "pipeline/cont_adhoc_sender.h"
+#include "pipeline/cont_adhoc_format.h"
+#include "libpq/pqformat.h"
+#include "libpq/libpq.h"
+#include "fmgr.h"
+#include "utils/lsyscache.h"
+#include "tcop/dest.h"
+
+/*
+ * Different key types to handle different types of queries.
+ *
+ * Append - select x::int from stream
+ * Single - select count(*) from stream
+ * Aggregate - select x::int, count(*) from stream group by x
+ */
+
+typedef enum
+{
+	Append,
+	Single,
+	Aggregate
+} KeyType;
+
+struct AdhocDestReceiver
+{
+	DestReceiver pub;
+
+	List *attnumlist;
+	FmgrInfo *out_funcs;
+	StringInfo msg_buf;
+	KeyType key_type;
+	unsigned row_count;
+
+	bool is_agg;
+	AttrNumber *keyColIdx;
+	int numCols;
+};
+
+/* send the buffered message in copy format, and flush.
+ * caller is responsible for clearing the buffer */
+
+static void
+adhoc_dest_send(struct AdhocDestReceiver *adhoc_dest)
+{
+	pq_putmessage('d', adhoc_dest->msg_buf->data, adhoc_dest->msg_buf->len);
```

Comment:
After reading through the logical decoding code, your decision to use a simple protocol wrapped in `COPY` makes a lot of sense. Logical decoding essentially provides a callback interface to do something with an `XLog` record when one is created. There's no standard textual representation as far as I can tell.

Actual continuous view updates will already work with logical decoding since they're WAL-logged, but it's worth thinking about how we can hook adhoc CQs in to the logical decoding infrastructure. It's not that important for the time being though.